### PR TITLE
join flow register user function, registration error component

### PR DESCRIPTION
**Pull request by benjiwheeler**
### Resolves:

Step towards resolving #3053

### Changes:

* add yesno checkbox
* allow checkbox to be hidden
* add registration error, waiting prop to join flow steps
* use redux to dispatch session refresh
* call register function in join flow
* make join flow modal stay showing even when session is refreshed

### Test Coverage:

none... are there tests we could add here?

### CODE REVIEW

File: src/components/join-flow/country-step.jsx [hunk 1]
```diff
@@ -94,6 +95,12 @@ class CountryStep extends React.Component {
                                     validate={this.validateSelect}
                                     validationClassName="validation-full-width-input"
                                 />
+                                <FormikCheckbox
```
[the remark this answers is not in the document]
Comment (benjiwheeler) [reply src/components/join-flow/country-step.jsx]: Done

File: src/components/join-flow/registration-error.jsx [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    constructor (props) {
+        super(props);
+        bindAll(this, [
+            'handleSubmit'
+        ]);
+        this.state = {
+        };
+    }
+    handleSubmit (e) {
+        e.preventDefault(); // prevent page reload
```
[the remark this answers is not in the document]
Comment (benjiwheeler) [reply src/components/join-flow/registration-error.jsx]: There's a <form> as part of the <JoinFlowStep>. As I understand it, by default, issuing a submit to a form will redirect the page.

File: src/components/join-flow/registration-error.jsx [hunk 3]
```diff
@@ -0,0 +1,44 @@
+const bindAll = require('lodash.bindall');
+const React = require('react');
+const PropTypes = require('prop-types');
+const {injectIntl, intlShape} = require('react-intl');
+
+const JoinFlowStep = require('./join-flow-step.jsx');
+
+require('./join-flow-steps.scss');
+
+class RegistrationError extends React.Component {
```
[the remark this answers is not in the document]
Comment (benjiwheeler) [reply src/components/join-flow/registration-error.jsx]: Sure, will do

File: src/components/join-flow/registration-error.jsx [hunk 4]
```diff
@@ -0,0 +1,44 @@
+const bindAll = require('lodash.bindall');
+const React = require('react');
+const PropTypes = require('prop-types');
+const {injectIntl, intlShape} = require('react-intl');
+
+const JoinFlowStep = require('./join-flow-step.jsx');
```
[the remark this answers is not in the document]
Comment (benjiwheeler) [reply src/components/join-flow/registration-error.jsx]: I don't know what that would look like! What is there to test here? I guess just that handleSubmit calls onTryAgain?

File: src/components/join-flow/registration-error.jsx [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+require('./join-flow-steps.scss');
+
+class RegistrationError extends React.Component {
+    constructor (props) {
+        super(props);
+        bindAll(this, [
+            'handleSubmit'
+        ]);
+        this.state = {
```
[the remark this answers is not in the document]
Comment (benjiwheeler) [reply src/components/join-flow/registration-error.jsx]: Yes, done

File: src/components/join-flow/join-flow.jsx [hunk 6]
```diff
@@ -21,48 +25,152 @@ class JoinFlow extends React.Component {
...
-    handleAdvanceStep (formData) {
-        formData = formData || {};
+    handleRegister (formData) {
+        this.setState({waiting: true}, () => {
+            api({
+                host: '',
+                uri: '/accounts/register_new_user/',
+                method: 'post',
+                useCsrf: true,
+                /* eslint-disable quote-props */
```
[the remark this answers is not in the document]
Comment (benjiwheeler) [reply src/components/join-flow/join-flow.jsx]: No big reason... just seemed to me like the rule doesn't really matter very much

File: src/components/join-flow/join-flow.jsx [hunk 7]
```diff
@@ -21,48 +25,152 @@ class JoinFlow extends React.Component {
...
+                    'g-recaptcha-response': null,
+                    gender: (
+                        formData.gender === 'other' ?
+                            formData.genderOther :
+                            formData.gender
+                    ),
+                    country: formData.country,
+                    subscribe: true,
+                    is_robot: formData.yesno
+                    // csrfmiddlewaretoken: 'abc'
```
[the remark this answers is not in the document]
Comment (benjiwheeler) [reply src/components/join-flow/join-flow.jsx]: Oh yeah, hmm, that was something I thought we had to submit with this request, but maybe we don't?

File: src/components/join-flow/join-flow.jsx [hunk 8]
```diff
@@ -21,48 +25,152 @@ class JoinFlow extends React.Component {
...
             formData: {},
             registrationError: null,
-            step: 0
+            step: 0,
+            waiting: false
         };
     }
-    handleAdvanceStep (formData) {
-        formData = formData || {};
+    handleRegister (formData) {
```
[the remark this answers is not in the document]
Comment (benjiwheeler) [reply src/components/join-flow/join-flow.jsx]: yeah, i don't really know how to do this!

File: src/components/join-flow/join-flow.jsx [hunk 9]
```diff
@@ -21,48 +25,152 @@ class JoinFlow extends React.Component {
...
+        });
+    }
+    handleAdvanceStep (newFormData) {
+        newFormData = newFormData || {};
+        const newState = {
+            formData: defaults({}, newFormData, this.state.formData)
+        };
+        // for the first 4 steps, automatically advance to next step.
+        // but for email step, we need to submit registration and wait.
+        const shouldAdvance = (this.state.step < 4);
```
[the remark this answers is not in the document]
Comment (benjiwheeler) [reply src/components/join-flow/join-flow.jsx]: Yeah, you're right. Revising.

File: src/components/join-flow/join-flow.jsx [hunk 10]
```diff
@@ -21,48 +25,152 @@ class JoinFlow extends React.Component {
...
+                        if (body && body[0]) {
+                            if (body[0].success) {
+                                this.props.refreshSession();
+                                this.setState({
+                                    step: this.state.step + 1
+                                });
+                                return;
+                            }
+                            if (body[0].errors) {
+                                const errorKeys = Object.keys(body[0].errors);
```
[the remark this answers is not in the document]
Comment (benjiwheeler) [reply src/components/join-flow/join-flow.jsx]: Added a comment.

**Comment by benjiwheeler** [comment]
Take a look at my revisions. We should talk about tests.

File: src/components/join-flow/join-flow.jsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (benjiwheeler) [reply src/components/join-flow/join-flow.jsx]: Trying with quotes

File: src/components/join-flow/join-flow.jsx [hunk 11]
```diff
@@ -21,48 +25,152 @@ class JoinFlow extends React.Component {
...
+                /* eslint-disable quote-props */
+                formData: {
+                    username: formData.username,
+                    email: formData.email,
+                    password: formData.password,
+                    birth_month: formData.birth_month,
+                    birth_year: formData.birth_year,
+                    'g-recaptcha-response': null,
+                    gender: (
+                        formData.gender === 'other' ?
```
[the remark this answers is not in the document]
Comment (benjiwheeler) [reply src/components/join-flow/join-flow.jsx]: This was not necessary, fixing!

File: src/components/join-flow/registration-error.jsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (benjiwheeler) [reply src/components/join-flow/registration-error.jsx]: Expanded comment

File: src/components/join-flow/join-flow.jsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (benjiwheeler) [reply src/components/join-flow/join-flow.jsx]: After discussion, we decided to take the anonymous function in handleSubmitRegistration and make it names, bind it to this, and pass it (err, body, res). Then we can write tests on this function in isolation.

File: src/components/join-flow/join-flow.jsx [hunk 12]
```diff
@@ -21,48 +25,151 @@ class JoinFlow extends React.Component {
...
-                    <UsernameStep onNextStep={this.handleAdvanceStep} />
-                    <BirthDateStep onNextStep={this.handleAdvanceStep} />
-                    <GenderStep onNextStep={this.handleAdvanceStep} />
-                    <CountryStep onNextStep={this.handleAdvanceStep} />
-                    <EmailStep onNextStep={this.handleAdvanceStep} />
-                    <WelcomeStep
-                        email={this.state.formData.email}
-                        username={this.state.formData.username}
-                        onNextStep={this.handleAdvanceStep}
+                {this.state.registrationError ? (
```
Comment (benjiwheeler) [inline src/components/join-flow/join-flow.jsx]: Write a test to confirm that if state.registrationerror exists, we show that and don't show progression. To do this, we'll need to shallow mount JoinFlow and call wrapper.setState().

**Comment by benjiwheeler** [comment]
Updated to address all issues, except tests!

Next: write tests for handleRegistrationResponse and RegistrationErrorStep.

**Comment by benjiwheeler** [comment]
Just one more test to add, and need to resolve conflicts.

**Comment by benjiwheeler** [comment]
added final test, and resolved conflicts

File: src/components/join-flow/join-flow.jsx [hunk 13]
```diff
@@ -13,56 +16,169 @@ const GenderStep = require('./gender-step.jsx');
...
+        //     "errors": {
+        //       "username": ["This field is required."],
+        //       "recaptcha": ["Incorrect, please try again."]
+        //     },
+        //     "success": false
+        //   }
+        // ]
+        this.setState({waiting: false}, () => {
+            let errStr = '';
+            if (!err && res.statusCode === 200) {
```
Comment (picklesrus) [inline src/components/join-flow/join-flow.jsx]: What about when `err` exists? we should handle timeouts, 403, 503, etc. but I don't think those cases are handled. As far as I can tell, we're only handling the case where we get a 200.

Comment (benjiwheeler) [reply src/components/join-flow/join-flow.jsx]: Yeah, in that case execution will continue after this block ends, and the next step will be to set `registrationError` and include the status code.

**Pull request merged**